### PR TITLE
*: implement the INSPECTION_SCHEMA to provide snapshot of inspection tables

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

### What problem does this PR solve? 

This PR introduces a new virtual system schema `inspection_schema`, which is used to provide a consistent view of `information_schema` tables, the related table will have the same table name within `information_schema`. The data will be obtained lazily from `information_schema` and cache in `SessionVars`, and the cached data will be cleared at `InspectionExec` closing.

This PR is part of #13567 

The reason why we should cache the data because some data of cluster-level memory tables will be retrieved many times in different inspection rules, and the cost of retrieving some data is expensive. We use the `TableSnapshot` to cache those data.

### What is changed and how it works?

1. Add a new struct `TableSnapshot`, which represents a data snapshot of the table contained in `inspection_schema`.
2. Add a new system schema `inspection_schema`.
3. Obtain data from `information_schema` lazily and cache them.

### Check List 

Tests 

 - Unit test

Release note

 - Introduces a new virtual system schema, which represents some inspection data snapshots and valid in the lifetime of `InspectionExec` executor lives.

### CODE REVIEW

File: util/misc.go [hunk 1]
```diff
@@ -143,6 +144,9 @@ const (
 	PerformanceSchemaLowerName = "performance_schema"
 )
 
+// InspectionSchemaName is the name of `inspection_schema`
+var InspectionSchemaName = model.NewCIStr("INSPECTION_SCHEMA")
```
Comment (djshow832) [inline util/misc.go]: Why not append it to L143?
Why is the table not added in `IsMemOrSysDB` below?

File: infoschema/inspection_schema_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	defer func() { c.Assert(failpoint.Disable(fpName), IsNil) }()
+
+	tk.MustQuery("select * from information_schema.cluster_info").Check(testkit.Rows(
+		"pd 127.0.0.1:11080 127.0.0.1:10080 mock-version mock-githash",
+		"tidb 127.0.0.1:11080 127.0.0.1:10080 mock-version mock-githash",
+		"tikv 127.0.0.1:11080 127.0.0.1:10080 mock-version mock-githash",
+	))
+
+	// enable inspection mode
+	inspectionTableCache := map[string]variable.TableSnapshot{}
```
Comment (djshow832) [inline infoschema/inspection_schema_test.go]: I didn't see the interface to enable it. So it's actually disabled and can't be enabled by user?

**Comment by zz-jason** [comment]
@lonng please add some proper labels.

File: infoschema/inspection_schema_test.go (hunk 2 above)
Comment (lonng) [reply infoschema/inspection_schema_test.go]: The `InspectionExec` will be implemented in #14114

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -484,6 +490,12 @@ type SessionVars struct {
 	MetricSchemaStep int64
 	// MetricSchemaRangeDuration indicates the step when query metric schema.
 	MetricSchemaRangeDuration int64
+
+	// Some data of cluster-level memory tables will be retrieved many times in different inspection rules,
+	// and the cost of retrieving some data is expensive. We use the `TableSnapshot` to cache those data
+	// and obtain them lazily, and provide a consistent view of inspection tables for each inspection rules.
+	// All cached snapshots will be released at the `InspectionExec` executor closing.
+	InspectionTableCache map[string]TableSnapshot
```
Comment (djshow832) [inline sessionctx/variable/session.go]: Do you mean that the snapshot is only cached in each session, and different sessions don't share the cache? Isn't it inefficient?

Comment (lonng) [reply sessionctx/variable/session.go]: @djshow832 Yes, each session may `SELECT * FROM inspection_result` at a different time and the result may be different. And the `InspectionTableCache` is a very short lives data.

File: meta/autoid/autoid.go [hunk 4]
```diff
@@ -42,6 +42,8 @@ const (
 	PerformanceSchemaDBID int64 = SystemSchemaIDFlag | 10000
 	// MetricSchemaDBID is the metric_schema schema id, it's exported for test.
 	MetricSchemaDBID int64 = SystemSchemaIDFlag | 20000
+	// InspectionSchemaDBID is the inspection_schema id, it's exports for test.
+	InspectionSchemaDBID int64 = SystemSchemaIDFlag | 30000
```
Comment (AilinKid) [inline meta/autoid/autoid.go]: Hard allocation code here should rebase autoID? if not, what if successive user creates schema/table with the same autoID.

File: infoschema/inspection_schema.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	// Obtain data from cache first
+	cached, found := sessionVars.InspectionTableCache[it.meta.Name.L]
+	if !found {
+		// We retrieve data from `information_schema` if can found in cache
+		rows, err := it.getRows(ctx, cols)
+		cached = variable.TableSnapshot{
+			Rows: rows,
+			Err:  err,
+		}
+		sessionVars.InspectionTableCache[it.meta.Name.L] = cached
```
[the remark this answers is not in the document]
Comment (lonng) [reply infoschema/inspection_schema.go]: The cached data will not be huge and will be released quickly (live duration inspection executor lives).

File: infoschema/inspection_schema.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	for name, cols := range inspectionTables {
+		tableInfo := buildTableMeta(name, cols)
+		tables = append(tables, tableInfo)
+		var ok bool
+		tid, ok := tableIDMap[tableInfo.Name.O]
+		if !ok {
+			panic(fmt.Sprintf("get inspection_schema table id failed, unknown system table `%v`", tableInfo.Name.O))
+		}
+		// Reuse information_schema table id serial number
+		tableInfo.ID = tid - autoid.InformationSchemaDBID + autoid.InspectionSchemaDBID
```
[the remark this answers is not in the document]
Comment (lonng) [reply infoschema/inspection_schema.go]: All tables in `information_schema` have a consecutive id like `autoid.InformationSchemaDBID + 1, autoid.InformationSchemaDBID + 2`. This line is use to replace `autoid.InformationSchemaDBID` with `autoid.InspectionSchemaDBID`.

File: meta/autoid/autoid.go (hunk 4 above)
Comment (lonng) [reply meta/autoid/autoid.go]: All id which allocated in `SystemSchemaIDFlag` namespace will not conflict with tables created by the user.

Comment (AilinKid) [reply meta/autoid/autoid.go]: even if the same ID is used both in `SystemSchemaIDFlag` and common table.

Comment (lonng) [reply meta/autoid/autoid.go]: Yes, the common table ID starts from 0 and will never conflict with the system table.

Comment (AilinKid) [reply meta/autoid/autoid.go]: See.

**Comment by AilinKid** [comment]
need fix the check dev and test.

**Pull request merged**